File: 백준/Silver/18870. 좌표 압축/좌표 압축.java
import java.util.*;
import java.io.*;
import java.util.stream.*;

class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        int N = Integer.parseInt(br.readLine());
        int[] points = Stream.of(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        int[] sortedPoints = Arrays.copyOf(points, N);
        Arrays.sort(sortedPoints);
        HashMap<Integer, Integer> map = new HashMap<>();
        
        int idx = 0;
        for(int i=0; i<N; i++) {
          if(i > 0 && sortedPoints[i] == sortedPoints[i - 1]) continue;
          map.put(sortedPoints[i], idx++);
        }
        
        StringBuilder sb = new StringBuilder();
        for(int i=0; i<N; i++) {
          sb.append(map.get(points[i]));
          if(i < N - 1) sb.append(" ");
        }
        sb.append("\n");
        
        bw.write(sb.toString());
        bw.flush();
        br.close();
        bw.close();
    }
}

File: 백준/Silver/18870. 좌표 압축/좌표 압축.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val n: Int = br.readLine().toInt()
    val points: List<Int> = br.readLine().split(" ").map { it.toInt() }
    val sortedPoints: List<Int> = points.sorted().distinct()
    val map = mutableMapOf<Int, Int>()
    for (i in sortedPoints.indices) {
        map[sortedPoints[i]] = i
    }
    val sb = StringBuilder()
    for(i in 0 until n) {
        sb.append(map[points[i]])
        if(i < n - 1) sb.append(" ")
    }
    sb.append("\n")
    BufferedWriter(OutputStreamWriter(System.`out`)).use {
        it.write(sb.toString())
        it.flush()
    }
    br.close()
}